src/lib.rs

```diff
@@ -40,19 +42,358 @@ mod repl_helper;
+                    machine.consult_module_string("facts", r_str.to_owned());
+                    0
+                },
+                Err(e_str) => {
+                    eprintln!("Error: {:?}", e_str);
+                    1
+                }
+            }
+        }
+
+        /// `scryer_run_query` runs a prolog query using a Prolog machine instance.
+        ///
+        /// This function accepts a mutable reference (`machine`) to an instance of the Prolog `Machine`
+        /// and an immutable C pointer (`input`) to a char which is to contain a valid Prolog query.
+        /// It attempts to run the query on the machine and return a serialized JSON string as a raw
+        /// C string pointer, with a "status" field and a "result" field.
+        /// If anything goes wrong, the "status" will be set to "error" and an "error" field is included.
+        /// However, if the operation is successful, the "status" is set to "ok" and the "result" field will contain the result of the query.
+        ///
+        /// # Safety
+        ///
+        /// This function contains unsafe blocks due to the use of raw pointers, but it is required
+        /// to operate across FFI boundaries. It is the caller's responsibility to ensure the validity
+        /// of `input`. The `input` should be a valid Null-Terminated `c_char` string.
+        ///
+        /// The returned C string is allocated on the heap and should be deallocated using the
+        /// `scryer_free_c_string()` function to prevent memory leaks.
+        ///
+        /// # Parameters
+        ///
+        /// - `machine`: a mutable reference to a `Machine` instance.
+        /// - `input`: an input in form of raw C `*const c_char` pointer.
+        ///
+        /// # Returns
+        ///
+        /// A raw pointer to a C string containing the serialized JSON result.
+        ///
+        /// # Expected Response Format
+        /// ```json
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": [ // Only present if status is "ok"
+        ///     { ... }, // Each is a Map representing a query result.
+        ///   ],
+        ///   "error": "..." // Only present if status is "error"
+        /// }
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_run_query(machine: &mut Machine, input: *const c_char) -> *mut c_char {
+            let c_string;
+            let r_str;
+            unsafe {
+                c_string = CStr::from_ptr(input);
+                r_str = c_string.to_str().expect("Not a valid UTF-8 string");
+            };
+            let result = std::panic::catch_unwind(|| r_str);
+
+            let output_string = match result {
+                Ok(r_str) => {
+                    let query_resolution = machine.run_query(r_str.to_owned());
+                    let value: serde_json::Value = serde_json::from_str(&format!("{}", query_resolution.expect("Something went wrong marshaling JSON"))).unwrap();
```

Comment:
You already have the possibility to handle errors so instead of calling `expect` on query_resolution the error should be handled properly.

Also instead of marshalling back and forth between rust and JSON
an `impl From<QueryResolution> for serde_json::Value` should probably be added. 
Alternatively, but more difficult a manual `serde::Serialize` implementation for `QueryResolution` would also do the trick.

Reply:
> Also instead of marshalling back and forth between rust and JSON
an `impl From<QueryResolution> for serde_json::Value` should probably be added.

This doesn't work because of the orphan rules, the maximum we could do is an `impl Into<serde_json::Value> for QueryResolution`. I think the best way to do that is to just implement a `to_json()` method.

Reply:
> > Also instead of marshalling back and forth between rust and JSON
> > an `impl From<QueryResolution> for serde_json::Value` should probably be added.
> 
> This doesn't work because of the orphan rules, the maximum we could do is an `impl Into<serde_json::Value> for QueryResolution`. I think the best way to do that is to just implement a `to_json()` method.

No, it actually doesn't violate coherence.
I tested adding 

```rust

impl From<QueryResolution> for serde_json::Value {
    fn from(value: QueryResolution) -> Self {
        todo!()
    }
}
```

just now and it passes the coherence check and compiles.

Reply:
`QueryResolution` (this is T1) is a local type and `serde_json::Value` (this is T0) doesn't contain any uncovered type parameters, so we are allowed to define this impl.

Reply:
Ok, [I didn't know this was allowed](https://stackoverflow.com/questions/75765502/rust-orphan-rule-and-from-trait).

Reply:
Another option is to assume that `QueryResolution::to_string` is valid json (its supposed to be as far as I can tell, but some edge cases are broken and I am currently working on a commit to fix this as part of #2460) and instead of using the `serde_json:json!` macro manually write that bit of json and just insert the result of `QueryResolution::to_string` at the right position.